### PR TITLE
Added check_snapshot option in the DB's AddFile function

### PR DESCRIPTION
**Pull request by baotiao**
This PR contain the changes that to add check_snapshot in DB's AddFile function

issue: #1249

### CODE REVIEW

File: include/rocksdb/db.h [hunk 1]
```diff
@@ -818,10 +818,10 @@ class DB {
   //        even if the file compression dont match the level compression
   virtual Status AddFile(ColumnFamilyHandle* column_family,
                          const std::vector<std::string>& file_path_list,
-                         bool move_file = false) = 0;
+                         bool move_file = false, bool check_snapshot = true) = 0;
```
Comment (IslamAbdelRahman) [inline include/rocksdb/db.h]: Let's rename `check_snapshot` to `skip_snapshot_check` and default it to false

File: include/rocksdb/db.h [hunk 2]
```diff
@@ -818,10 +818,10 @@ class DB {
   //        even if the file compression dont match the level compression
```
Comment (IslamAbdelRahman) [inline include/rocksdb/db.h]: Let's update the comments of the function to include the new option, and the implications of using it

**Comment by IslamAbdelRahman** [comment]
Thanks @baotiao, Let's also add a unit test that uses this new option and make sure that when we use it, snapshots see the data that is added in the new file

**Comment by dhruba** [comment]
Just curious to understand the requirement of skipping the snapshot check, is it because the ,snapshot, check is very heavyweight in performance?

**Comment by IslamAbdelRahman** [comment]
@dhruba, Actually the snapshot check is very lightweight. This feature is for people who want to use the AddFile API while holding snapshots (#1249) and don't mind the implications of doing that.

The implication of ingesting a file while holding a snapshot is that if you took a snapshot of the DB and then ingest a file, the data in this file will be visible in this snapshot even though AddFile() was called after the snapshot was taken (because data seqno in the file is 0).

**Comment by dhruba** [comment]
Thanks for the description @IslamAbdelRahman

**Comment by baotiao** [comment]
@IslamAbdelRahman  I have rename the variable, change the comment and add the corresponding unit test

File: include/rocksdb/db.h [hunk 3]
```diff
@@ -806,22 +806,23 @@ class DB {
...
-  // move the files instead of copying them.
+  // move the files instead of copying them, skip_snapshot_check can be set to
+  // true to ignore the snapshot, make sure that you know that when you use it,
+  // snapshots see the data that is added in the new files.
   //
   // Current Requirements:
   // (1) The key ranges of the files don't overlap with each other
   // (2) The key range of any file in list doesn't overlap with
   //     existing keys or tombstones in DB.
-  // (3) No snapshots are held.
```
Comment (IslamAbdelRahman) [inline include/rocksdb/db.h]: let's keep this line but update it to
`(3) No snapshots are held (check skip_snapshot_check to skip this check)`.

**Comment by IslamAbdelRahman** [comment]
Thanks @baotiao, Looks great .. the test failures look unrelated
I have one final minor comment

**Comment by baotiao** [comment]
Thanks @IslamAbdelRahman. the failures test show that if we holding snapshot, the Addfile will failed when we use the default skip_snapshot_check.
I have changed the comment.

**Pull request merged**